File: section11/eazystore/src/main/java/com/eazybytes/eazystore/exception/GlobalExceptionHandler.java
package com.eazybytes.eazystore.exception;

import java.time.LocalDateTime;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;

import com.eazybytes.eazystore.dto.ErrorResponseDto;

@RestControllerAdvice
public class GlobalExceptionHandler {

    @ExceptionHandler(Exception.class)
    public ResponseEntity<ErrorResponseDto> handleGlobalException(Exception exception, WebRequest webRequest) {
        ErrorResponseDto errorResponseDto = new ErrorResponseDto(webRequest.getDescription(false),
                HttpStatus.INTERNAL_SERVER_ERROR, exception.getMessage(), LocalDateTime.now());

        // return ResponseEntity.internalServerError().body(errorResponseDto);
        return new ResponseEntity<>(errorResponseDto, HttpStatus.INTERNAL_SERVER_ERROR);
    }
}

File: backend/courses/inflearn/kotlin-spring-api-practice/src/main/kotlin/com/example/blog/core/exception/GlobalExceptionHandler.kt
package com.example.blog.core.exception

import com.example.blog.core.response.ErrorResponse
import org.slf4j.LoggerFactory
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class GlobalExceptionHandler {

    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * Handles custom business exceptions.
     */
    @ExceptionHandler(BusinessException::class)
    fun handleBusinessException(e: BusinessException): ResponseEntity<ErrorResponse> {
        log.warn("BusinessException: {}", e.message)
        val errorCode = e.errorCode
        val response = ErrorResponse(
            code = errorCode.code,
            message = e.message ?: errorCode.message
        )
        return ResponseEntity(response, errorCode.status)
    }

    /**
     * Handles DTO validation failures (@Valid).
     */
    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleMethodArgumentNotValidException(e: MethodArgumentNotValidException): ResponseEntity<ErrorResponse> {
        log.warn("MethodArgumentNotValidException: {}", e.message)
        val errorCode = ErrorCode.INVALID_INPUT_VALUE
        val response = ErrorResponse(
            code = errorCode.code,
            message = errorCode.message,
            fieldErrors = ErrorResponse.FieldError.of(e.bindingResult)
        )
        return ResponseEntity(response, errorCode.status)
    }

    /**
     * Fallback handler for all unexpected exceptions.
     */
    @ExceptionHandler(Exception::class)
    fun handleException(e: Exception): ResponseEntity<ErrorResponse> {
        log.error("Unhandled Exception", e)
        val errorCode = ErrorCode.INTERNAL_SERVER_ERROR
        val response = ErrorResponse(
            code = errorCode.code,
            message = errorCode.message
        )
        return ResponseEntity(response, errorCode.status)
    }
}

